File: 74. Search a 2D Matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        B, R = len(matrix), len(matrix[0])
        T = 0
        while T < B:
            if target >= matrix[T][0] and target <= matrix[T][R - 1]:
                L = 0
                while L < R:
                    if matrix[T][L] == target:
                        return True
                    L += 1
            T += 1

File: 74. Search a 2D Matrix.c


bool searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target){
    
    int totalSize = matrixSize * (*matrixColSize);
    int left = 0, right = totalSize - 1;
    while (left <= right){
        int mid = (left + right) / 2;
        int row = mid / (*matrixColSize);
        int col = mid % (*matrixColSize);
        if (matrix[row][col] == target){
            return true;
        }else if (matrix[row][col] > target){
            right = mid - 1;
        }else{
            left = mid + 1;
        }
    }
    return false;
}
